pkg/controller/machinepool/openstackactuator.go

```diff
@@ -139,37 +138,30 @@ func (a *OpenStackActuator) GenerateMachineSets(cd *hivev1.ClusterDeployment, po
 }
 
 // Get the OS image from an existing master machine.
-func getOpenStackOSImage(masterMachine *machinev1beta1.Machine, scheme *runtime.Scheme, logger log.FieldLogger) (string, error) {
-	providerSpec, err := decodeOpenStackMachineProviderSpec(masterMachine.Spec.ProviderSpec.Value, scheme)
-	if err != nil {
+func getOpenStackOSImage(masterMachine *machinev1beta1.Machine, logger log.FieldLogger) (string, error) {
+	var u unstructured.Unstructured
+	if err := json.Unmarshal(masterMachine.Spec.ProviderSpec.Value.Raw, &u); err != nil {
 		logger.WithError(err).Warn("cannot decode OpenstackProviderSpec from master machine")
-		return "", errors.Wrap(err, "cannot decode OpenstackProviderSpec from master machine")
-	}
-	var osImage string
-	if providerSpec.RootVolume != nil {
-		osImage = providerSpec.RootVolume.SourceUUID
-	} else {
-		osImage = providerSpec.Image
+		return "", errors.Wrap(err, "Failed to unmarshal master machine OpenstackProviderSpec to Unstructured")
 	}
+	osImage := osImageFromUnstructuredProviderSpec(&u)
 	logger.WithField("image", osImage).Debug("resolved image to use for new machinesets")
 	return osImage, nil
 }
 
-func decodeOpenStackMachineProviderSpec(rawExt *runtime.RawExtension, scheme *runtime.Scheme) (*machinev1alpha1.OpenstackProviderSpec, error) {
-	codecFactory := serializer.NewCodecFactory(scheme)
-	decoder := codecFactory.UniversalDecoder(machinev1alpha1.GroupVersion)
-	if rawExt == nil {
-		return nil, fmt.Errorf("MachineSet has no ProviderSpec")
-	}
-	obj, gvk, err := decoder.Decode([]byte(rawExt.Raw), nil, nil)
-	if err != nil {
-		return nil, fmt.Errorf("could not decode OpenStack ProviderSpec: %v", err)
-	}
-	spec, ok := obj.(*machinev1alpha1.OpenstackProviderSpec)
-	if !ok {
-		return nil, fmt.Errorf("unexpected object: %#v", gvk)
+// osImageFromUnstructuredProviderSpec exists because it's hard to decode OpenstackProviderSpec into a go object
+// because, depending on the version of the cluster from which it was obtained, it may be one of two different
+// apiVersions. Rather than trying to figure out which one, just path into it in its Unstructured form.
+func osImageFromUnstructuredProviderSpec(u *unstructured.Unstructured) string {
+	rvi, ok := u.Object["rootVolume"]
+	if ok {
+		rv := rvi.(map[string]interface{})
+		sourceUUID, ok := rv["sourceUUID"]
+		if ok && sourceUUID != "" {
+			return sourceUUID.(string)
+		}
 	}
-	return spec, nil
+	return u.Object["image"].(string)
```

Comment:
Would it be good practice to add some error handling here, in case of an unexpected object error (i.e. future api changes such as this one)?

Reply:
Totally. I'll work something up.

Reply:
I'm wondering if it is possible to use utilities defined in the unstructured package like `unstructured.NestedString` ? 
e.g. https://github.com/openshift/hive/blob/25794ae6cc5e98e9cb65f84c8fba8666bcfe1070/vendor/k8s.io/kubectl/pkg/cmd/wait/wait.go#L536-L564

Reply:
I also considered using gjson. Figured the paths were small/simple enough that it wasn't worth it... but now looking at it I'm not so sure. Let me play...

Reply:
Yeah, I like that a lot better @fxierh, changed. Thanks for the suggestion!